File: 프로그래머스/lv2/17684. ［3차］ 압축/［3차］ 압축.js
function solution(msg) {
    const alphabet = "abcdefghijklmnopqrstuvwxyz".toUpperCase();
    const container = [...alphabet].reduce((accumulator, value, index) => {
        return { ...accumulator, [value]: ++index };
    }, {});

    const answer = [];
    while (msg.length) {
        let i, w;
        for (i = msg.length; i > 0; --i) {
            w = msg.slice(0, i);
            if (container.hasOwnProperty(w)) {
                answer.push(container[w]);
                break;
            }
        }
        msg = msg.slice(i);
        container[w + msg[0]] = Object.keys(container).length + 1
    }

    return answer;
}

File: 프로그래머스/lv2/17684. ［3차］ 압축/［3차］ 압축.cpp
#include <string>
#include <vector>
#include <unordered_map>

using namespace std;

#define ALPHABET 64

vector<int> solution(string msg) {
    vector<int> answer;
    unordered_map<string, int> m;

    for (size_t i = 0; i < msg.size();)
    {
        int j = 2;
        while (m.count(msg.substr(i, j)) && i + j <= msg.size()) ++j;

        if (j == 2) answer.emplace_back(msg[i] - ALPHABET);
        else answer.emplace_back(m[msg.substr(i, j - 1)]);
        
        m.insert(make_pair(msg.substr(i, j), 27 + m.size()));
        i += j - 1;
    }

    return answer;
}